src/main/kotlin/org/rust/ide/inspections/RsConstReferStaticInspection.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.psi.util.forEachDescendantOfType
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.isConst
+import org.rust.lang.utils.RsDiagnostic
+import org.rust.lang.utils.addToHolder
+
+class RsConstReferStaticInspection : RsLocalInspectionTool() {
+
+    override fun buildVisitor(holder: RsProblemsHolder, isOnTheFly: Boolean) = object : RsVisitor() {
+        override fun visitConstant(constant: RsConstant) {
+            // We only want to target `const` constants (thus, excluding `static`) with this inspection.
+            if (!constant.isConst) return
+
+            errorOnStaticRef(constant.expr ?: return, holder) { expr ->
+                RsDiagnostic.ConstVariableReferToStaticError(expr, constant, expr.text)
+            }
+        }
+
+        override fun visitArrayType(arrayType: RsArrayType) {
+            errorOnStaticRef(arrayType.expr ?: return, holder) { expr ->
+                RsDiagnostic.ConstArraySizeReferToStaticError(expr, expr.text)
+            }
+        }
+
+        override fun visitArrayExpr(arrayExpr: RsArrayExpr) {
+            val lastExpr = arrayExpr.exprList.lastOrNull()
+            if (arrayExpr.semicolon != null && lastExpr != null) {
+                errorOnStaticRef(lastExpr, holder) { expr ->
+                    RsDiagnostic.ConstArraySizeReferToStaticError(expr, expr.text)
+                }
+            }
+        }
+
+        override fun visitEnumVariant(enumVariant: RsEnumVariant) {
+            val name = enumVariant.name ?: return
+            errorOnStaticRef(enumVariant.variantDiscriminant?.expr ?: return, holder) { expr ->
+                RsDiagnostic.ConstEnumDiscriminantReferToStaticError(expr, name, expr.text)
+            }
+        }
+
+        override fun visitCallExpr(call: RsCallExpr) {
+            val typeArgumentList = (call.expr as? RsPathExpr ?: return).path.typeArgumentList ?: return
+            val constParams = ((call.expr as RsPathExpr).path.reference?.resolve() as RsFunction).typeParameterList?.constParameterList
+            typeArgumentList.exprList.forEach {
+                // TODO: Find out which const param this expression targets, and skip if it is not const.
```

Comment:
I'm not sure how to do this.
Consider the following:
```rust
static X: usize = 5;

fn foo<A, const C: usize>() {}

fn main() {
    foo::<bool, X>();
}
```
in `foo::<bool, X>`, `X` is the second element in `typeArgumentList`. 
However, it matches the first entry in `constParams`. How do I reliably figure out whether or not this specific expression is supposed to be `const`?
Is the order always assured to be the same?

Reply:
Sorry for the long response!

For now, we parse `X` in `foo::<bool, X>()` as a type reference, so without fixing the grammar we cannot handle this case. You can try to support `foo::<bool, { X }>()` / `foo::<{ X }, bool>()` cases - here `{ X }` is parsed as an expression, so it will be the first in the `constArguments` (declared in`RsPath.kt`) in both cases.